api/envoy/data/dns/v3/dns_table.proto

```diff
@@ -39,18 +39,74 @@ message DnsTable {
     }];
   }
 
-  // This message type is extensible and can contain a list of addresses, clusters or
-  // dictate a different method for resolving the addresses for an endpoint
+  // Specify the service protocol using a numeric or string value
+  message DnsServiceProtocol {
+    oneof protocol_config {
+      option (validate.required) = true;
+
+      // Specify the protocol number for the service. Envoy will try to resolve the number to
+      // the protocol name. For example, 6 will resolve to "tcp". Refer to:
+      // https://www.iana.org/assignments/protocol-numbers/protocol-numbers.xhtml
+      // for protocol names and numbers
+      uint32 number = 1 [(validate.rules).uint32 = {lt: 255}];
+
+      // Specify the protocol name for the service.
+      string name = 2 [(validate.rules).string = {min_len: 1 well_known_regex: HTTP_HEADER_NAME}];
+    }
+  }
+
+  // Specify the target for a given DNS service
+  message DnsServiceTarget {
+    // Use a resolvable hostname or cluster as the endpoint for a service. Whether the name is
+    // a host or a cluster we will return all resolved addresses in the answer to the client
+    string name = 1 [(validate.rules).string = {min_len: 1 well_known_regex: HTTP_HEADER_NAME}];
+
+    // The priority of the service record target
+    uint32 priority = 2 [(validate.rules).uint32 = {lt: 65536}];
+
+    // The weight of the service record target
+    uint32 weight = 3 [(validate.rules).uint32 = {lt: 65536}];
+  }
+
+  // This message defines a service selection record returned for a service query in a domain
+  // [#next-free-field: 6]
+  message DnsService {
+    // The name of the service without the protocol or domain name
+    string service_name = 1;
+
+    // The service protocol. This can be specified as a string or the numeric value of the protocol
+    DnsServiceProtocol protocol = 2;
+
+    // The service entry time to live. This is independent from the DNS Answer record TTL
+    google.protobuf.Duration ttl = 3 [(validate.rules).duration = {gte {seconds: 1}}];
+
+    // The port to which the service is bound
```

Comment:
If the DnsServiceTarget has a cluster name, wouldn't the port be retrieved from there? Might be useful to move the port inside the DnsServiceTarget, where there is a one of { HostName and Port OR Cluster Name }

Reply:
Let me work on this aspect and see what I can do.

Reply:
Added an integration test that verifies this.